### DISCUSSION TITLE
publishing different timestamp

### DISCUSSION THREAD

**Discussion by sfalahati** [grafana/mimir#4292] [General]
Hi Team,

I've been following discussion 4009 to be able to push data to mimir (as opposed to pull data which I understand is the default behavior).
My scenario is that I have already captured my monitoring metrics using a different system and now only need to store them in mimir and visualize them. Now, if I write a program that reads the data from third party database and push them (request.post) to mimir url, can I manipulate the time. I need to graph the data as the time they were captured in the other system not the time that they are pushed to mimir.

So in below line:

```
sample_result = mimir_pb2.Sample(timestamp_ms=round(time.time() * 1000), value=1)
timeseries_result = mimir_pb2.TimeSeries(labels=[label_pair], samples=[sample_result])
```

can I read the timestamp from third party and use that to push data:
```
sample_result = mimir_pb2.Sample(timestamp_ms=<Value in 3rd party tool>, value=<Value stored in 3rd party)
timeseries_result = mimir_pb2.TimeSeries(labels=[label_pair], samples=[sample_result])
```

I tried to experiment this. Pushing data with current timestamp (timestamp_ms=round(time.time() * 1000)) show the graph as of the timestamp generated. But when I manipulate the time and write another timestamp, it doesn't show the data. (ButI receive 200 http status code that data is successfully pushed to mimir).

**Comment by pstibrany** [comment]
> But when I manipulate the time and write another timestamp, it doesn't show the data. (ButI receive 200 http status code that data is successfully pushed to mimir).

If you have got 200, it means that ingesters have accepted the data. What timestamp (relative to "now") are you using?

You may be running into problem with caching -- Mimir normally caches results of queries for timestamps older than 10m. Perhaps the empty result you're seeing is actually a cached result?

If you use very old timestamps (older than 12h), another possibility is that queriers don't even ask ingesters for that data, and you will need to extend `query-ingesters-within` parameter.

**Reply by sfalahati** [reply]
Thanks Peter. Is the parameter 'query-ingesters-within' for loki or Mimir? (We are using Grafana Mimir).

Also, do you see any issue if data is pushed out of synch? I pushed some data successfully and could see them in our dashboard. Then I manipulated the time and pushed some data for a day before (built timestamp for yesterday). The push was successful, but data was not visualizing. I don't know if it's because perhaps the new data had timestamp older than existing data which was already being visualised or there is another reason for it.

**Reply by pstibrany** [reply]
> Thanks Peter. Is the parameter 'query-ingesters-within' for loki or Mimir? (We are using Grafana Mimir).

Full CLI option to set this flag is `-querier.query-ingesters-within`.

> The push was successful, but data was not visualizing. I don't know if it's because perhaps the new data had timestamp older than existing data which was already being visualised or there is another reason for it.

If your `-querier.query-ingesters-within` is correct (it needs to cover timerange of your newly pushed data), this can be a caching issue.

**Comment by sfalahati** [comment]
Hi Mimir team, following up on this post, how far back can we go for pushing data. I am getting below error when run a request.post trying to push data that is 24 hours old:

_failed pushing to ingester: user=<USER>: the sample has been rejected because its timestamp is too old_ 

I tried various timestamp and I cannot push any metric older than 60 minutes. This is as far as I can go back to be able to successfully write data to Mimir. Is this setting configurable? 

My architecture is that I have a system that already has the metrics in its database. I need to read them and write them to Mimir to visualize with the timestamp coming from the other system. Now if the job that reads metric data from other system and write the data to Mimir doesn't work for few hours, I am concern that I cannot write the older data to Mimir once the job starts again.
